lib/auth/auth_with_roles.go

```diff
@@ -765,6 +765,47 @@ func (a *ServerWithRoles) GetNodes(ctx context.Context, namespace string, opts .
 	return filteredNodes, nil
 }
 
+// ListResources returns a paginated list of resources filtered by user access.
+func (a *ServerWithRoles) ListResources(ctx context.Context, req proto.ListResourcesRequest) ([]types.Resource, string, error) {
+	limit := int(req.Limit)
+	if limit <= 0 {
+		return nil, "", trace.BadParameter("nonpositive parameter limit")
+	}
+
+	var resourceKind string
+	var checkAccessFunc func(types.Resource) error
+	switch req.ResourceType {
+	case proto.ResourceType_RESOURCE_TYPE_DATABASE_SERVER:
+		resourceKind = types.KindDatabaseServer
+		checkAccessFunc = a.checkAccessToDatabaseServer
+	case proto.ResourceType_RESOURCE_TYPE_APPLICATION_SERVER:
+		resourceKind = types.KindAppServer
+		checkAccessFunc = a.checkAccessToAppServer
+	default:
+		return nil, "", trace.Errorf("unsupported resource type %s", req.ResourceType)
+	}
+
+	if err := a.action(req.Namespace, resourceKind, types.VerbList); err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	var resources []types.Resource
+	nextKey, err := a.authServer.IterateResourcePages(ctx, req, func(nextPage []types.Resource) (bool, error) {
+		for _, resource := range nextPage {
+			if err := checkAccessFunc(resource); err == nil {
+				resources = append(resources, resource)
```

Comment:
Are we really want ignore error ? I think that skip should be done only on !trace.IsAccessDenied(err) otherwise the err should be returned.

Reply:
It makes sense. I've changed it to return the error when it is not an `AccessDenied`.